freeipa/src/main/resources/freeipa-salt/salt/monitoring/scripts/cipa.sh

```diff
@@ -0,0 +1,103 @@
+#!/bin/bash
+
+function parse_table() {
+  # first awk: split table by | char from the second line and skip lines that starts with and print columns as comma separated values +--
+  # last awk: the first column (comma separated) -> lowercase + replace space with underscore
+  # output: comma separated fields based on thw formatted table
+  cat /tmp/cipa_out.txt | awk -F'|' ' NR > 1 && !/^+--/ {printf $2;for(i=3;i<NF;++i)printf ","$i; printf "\n"}' | awk '{$1=$1};1' | sed 's/\s*,\s*/,/g' | grep -v '^,*$' | awk  -F',' '{l=tolower($1); un=gsub(/ /,"_",l); printf l; for(i=2;i<=NF;++i)printf ","$i; printf "\n" }'
```

Comment:
It's a miracle that it works but can't we have some python script which does the same and easier to understand? I mean instead of the whole script.

Reply:
I check if this is possible. cipa internally uses https://pypi.org/project/prettytable/ which can print the table as json with `table.get_json_string()`.

The solution is to patch the cipa python script `/usr/local/lib/python3.6/site-packages/checkipaconsistency/main.py` and replace this line https://github.com/peterpakos/checkipaconsistency/blob/master/checkipaconsistency/main.py#L268

```python
        self._log.info(table)
```

with this

```python
        self._log.info(table.get_json_string())
```

Then we call cipa like this

```bash
cipa --no-header
```

The result is:

```json
[
    {
        "FreeIPA servers:": "Active Users",
        "STATE": "OK",
        "ipaserver0": "102",
        "ipaserver1": "102"
    },
    {
        "FreeIPA servers:": "Stage Users",
        "STATE": "OK",
        "ipaserver0": "0",
        "ipaserver1": "0"
    },
    {
        "FreeIPA servers:": "Preserved Users",
        "STATE": "OK",
        "ipaserver0": "0",
        "ipaserver1": "0"
    },
    {
        "FreeIPA servers:": "Hosts",
        "STATE": "OK",
        "ipaserver0": 6,
        "ipaserver1": 6
    },
    {
        "FreeIPA servers:": "Services",
        "STATE": "OK",
        "ipaserver0": 14,
        "ipaserver1": 14
    },
    {
        "FreeIPA servers:": "User Groups",
        "STATE": "OK",
        "ipaserver0": 29,
        "ipaserver1": 29
    },
    {
        "FreeIPA servers:": "Host Groups",
        "STATE": "OK",
        "ipaserver0": "1",
        "ipaserver1": "1"
    },
    {
        "FreeIPA servers:": "Netgroups",
        "STATE": "OK",
        "ipaserver0": 0,
        "ipaserver1": 0
    },
    {
        "FreeIPA servers:": "HBAC Rules",
        "STATE": "OK",
        "ipaserver0": 2,
        "ipaserver1": 2
    },
    {
        "FreeIPA servers:": "SUDO Rules",
        "STATE": "OK",
        "ipaserver0": 0,
        "ipaserver1": 0
    },
    {
        "FreeIPA servers:": "DNS Zones",
        "STATE": "OK",
        "ipaserver0": 3,
        "ipaserver1": 3
    },
    {
        "FreeIPA servers:": "Certificates",
        "STATE": "OK",
        "ipaserver0": 18,
        "ipaserver1": 18
    },
    {
        "FreeIPA servers:": "LDAP Conflicts",
        "STATE": "OK",
        "ipaserver0": 0,
        "ipaserver1": 0
    },
    {
        "FreeIPA servers:": "Ghost Replicas",
        "STATE": "OK",
        "ipaserver0": 0,
        "ipaserver1": 0
    },
    {
        "FreeIPA servers:": "Anonymous BIND",
        "STATE": "OK",
        "ipaserver0": "ROOTDSE",
        "ipaserver1": "ROOTDSE"
    },
    {
        "FreeIPA servers:": "Microsoft ADTrust",
        "STATE": "OK",
        "ipaserver0": false,
        "ipaserver1": false
    },
    {
        "FreeIPA servers:": "Replication Status",
        "STATE": "OK",
        "ipaserver0": "ipaserver1 0",
        "ipaserver1": "ipaserver0 0"
    }
]
```

Compared to the original table output:

```
+--------------------+--------------+--------------+-------+
| FreeIPA servers:   | ipaserver0   | ipaserver1   | STATE |
+--------------------+--------------+--------------+-------+
| Active Users       | 102          | 102          | OK    |
| Stage Users        | 0            | 0            | OK    |
| Preserved Users    | 0            | 0            | OK    |
| Hosts              | 6            | 6            | OK    |
| Services           | 14           | 14           | OK    |
| User Groups        | 29           | 29           | OK    |
| Host Groups        | 1            | 1            | OK    |
| Netgroups          | 0            | 0            | OK    |
| HBAC Rules         | 2            | 2            | OK    |
| SUDO Rules         | 0            | 0            | OK    |
| DNS Zones          | 3            | 3            | OK    |
| Certificates       | 18           | 18           | OK    |
| LDAP Conflicts     | 0            | 0            | OK    |
| Ghost Replicas     | 0            | 0            | OK    |
| Anonymous BIND     | ROOTDSE      | ROOTDSE      | OK    |
| Microsoft ADTrust  | False        | False        | OK    |
| Replication Status | ipaserver1 0 | ipaserver0 0 | OK    |
+--------------------+--------------+--------------+-------+
```

Reply:
I agree, JSON format would be much better.

Reply:
that project is GPLv3, as my understanding that will require to add any state changes to the copyright notice, but not clear then should be add it here (if dynamically patch the code) or modify it in cloudbreak-images

Reply:
what i can offer for now is to have a python implementation if that's more readable for others (which works by both python2/python3), otherwise if we want to change the cipa code, these are the options 
a.) update python code for cipa in cb or cb images - will require to change license to gpl - i guess we do not want that
b) fork cipa and add json format option there, put the project into hortonworks github, and make sure that is public + contain state changes + have gpl version (will need to check with legal team as well)
c.) instead of installing cipa from pip, use a custom forked version that has json output (i guess there are already other solutions for that as well) and burn that one to cb images (not the original one) - i guess that would not be installed by pip, so we will need the python files + symlink + license+state changes description

Reply:
or just do a) without license changes
or create a pr for the original project, and get a release, maybe it's really fast

Reply:
not sure about a.), even if it's an on-the-fly change that could impact license (?)
PR could take long time as there are still some open PRs there (for some time)

Reply:
+1 for creating a PR. I think this would be really useful for others as well and the change is just specifying some command line flag to report in json and using it when it logs to the console.

Reply:
@foldik , there is a PR already in that project (adding json mode), that is open for 2 years now, so I would not count with that